data/lib/core/position.lua

```diff
@@ -35,7 +35,7 @@ function Position:moveUpstairs()
 				direction = DIRECTION_WEST
 			end
 
-			local position = self + Position.directionOffset[direction]
+			local position = Position(self.x + Position.directionOffset[direction].x, self.y + Position.directionOffset[direction].y, self.z)
```

Comment:
`Position:getNextPosition(direction)` already does that, so

```lua
local position = self:getNextPosition(direction)
```

should have the same result, but cleaner

Reply:
This will not work by the looks of it. It seems that Position:getNextPosition(direction) applies the offsets to itself, so on the next iteration in the loop, it will then be applying the next offsets to that and not the original position.

Reply:
I don't think it's ideal behavior.
![image](https://user-images.githubusercontent.com/26801045/220474038-45336b0c-465f-40af-89e0-7c59c003b2ea.png)

Reply:
> I don't think it's ideal behavior. ![image](https://user-images.githubusercontent.com/26801045/220474038-45336b0c-465f-40af-89e0-7c59c003b2ea.png)

Your :isWalkable() method is clearly busted. Or that table item doesn't have TILESTATE_BLOCKSOLID or TILESTATE_IMMOVABLEBLOCKSOLID.

Reply:
> It seems that Position:getNextPosition(direction) applies the offsets to itself

Ah this awful behavior. But you can clone the position:
```lua
local position = Position(self)
position:getNextPosition(direction)
```

Reply:
> > It seems that Position:getNextPosition(direction) applies the offsets to itself
> 
> Ah this awful behavior. But you can clone the position:
> 
> ```lua
> local position = Position(self)
> position:getNextPosition(direction)
> ```

works fine